### ISSUE TITLE
fresh build, page loads but working icon in the middle never finishes

### ISSUE THREAD

**Issue by digitalw00t** [ItzCrazyKns/Perplexica#62]
**Describe the bug**
I did a fresh docker build, got to localhost:3000, and the page never finishes loading.  I get the icons on the left side, and a swirling circle in the middle that never finishes.

**To Reproduce**
Same as in description

**Expected behavior**
Page to load and be able to do a search.

This was from a docker compose up, so I would have assumed it was going to be a quick install and startup.

**Comment by digitalw00t** [comment]
Played around, got it to let me load the settings page, and I'm getting:
perplexica-backend-1   | error: Error loading Ollama models: TypeError: fetch failed
perplexica-backend-1   | error: Error loading Ollama embeddings: TypeError: fetch failed

When I use the pulldown selection for the models, they are empty.  Ollama is already running on this host in a service and is working and has been tested.  It can be accessed via localhost or remotely from the network by another machine.

**Comment by DaslavCl** [comment]
doesnt run with any kind of config the page never finishes loading

**Comment by apulache** [comment]
It seems it get stuck in the loading models stage

![Screenshot from 2024-05-04 20-09-16](https://github.com/ItzCrazyKns/Perplexica/assets/2507738/5b4987cf-13d7-499d-8afe-5fbe1f5908d1)

Please review it.

Thanks in advance!

**Comment by ItzCrazyKns** [comment]
Can everyone paste their logs from the backend Docker container?

**Comment by ItzCrazyKns** [comment]
> Played around, got it to let me load the settings page, and I'm getting: perplexica-backend-1 | error: Error loading Ollama models: TypeError: fetch failed perplexica-backend-1 | error: Error loading Ollama embeddings: TypeError: fetch failed
> 
> When I use the pulldown selection for the models, they are empty. Ollama is already running on this host in a service and is working and has been tested. It can be accessed via localhost or remotely from the network by another machine.

It suggests that the backend is not able to access Ollama's API. Make sure you're using the correct IP/Address in the config for Ollama's API URL

**Comment by DaslavCl** [comment]
Same issue setting OpenAI

**Comment by ItzCrazyKns** [comment]
> Played around, got it to let me load the settings page, and I'm getting: perplexica-backend-1 | error: Error loading Ollama models: TypeError: fetch failed perplexica-backend-1 | error: Error loading Ollama embeddings: TypeError: fetch failed
> 
> When I use the pulldown selection for the models, they are empty. Ollama is already running on this host in a service and is working and has been tested. It can be accessed via localhost or remotely from the network by another machine.

Localhost for the Docker container would be for its network and not from the host's network. That's why we make use of host.docker.internal thought its only available in Docker desktops, Linux users still face issues.

**Comment by cpedia** [comment]
Even a fresh deployment to RepoCloud doesn't work.

**Comment by rmensing** [comment]
Windows - docker-desktop
host ip address: 192.168.1.3
Fresh install and a fresh run.
Ollama running on host using `ollama serve`
Confirmed working using Open-webui running in docker - connects to ollama api with: http://192.168.1.3:11434
frontend ui has spinner
frontend settings does not show ollama

config.toml:
```
[API_KEYS]
OPENAI = ""
GROQ = ""

[API_ENDPOINTS]
OLLAMA = "http://192.168.1.3:11434"
SEARXNG = "http://localhost:32768"

[GENERAL]
PORT = 3001
SIMILARITY_MEASURE = "cosine"
```
backend logs:
```
2024-05-05 04:45:15 (node:28) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
2024-05-05 04:45:15 (Use `node --trace-deprecation ...` to show where the warning was created)
2024-05-05 04:45:15 yarn run v1.22.19
2024-05-05 04:45:15 $ node dist/app.js
2024-05-05 04:45:15 info: WebSocket server started on port 3001
2024-05-05 04:45:15 info: Server is running on port 3001
```
ollama is accessible from the container using the host IP:
(install curl in backend container: `apk update; apk add curl`)
```
/home/perplexica # curl http://192.168.1.3:11434/
Ollama is running
```

Also tried using `host.docker.internal` in config.toml and testing with curl. does the same thing.

The only activity the ollama log shows is the curl requests, no connection attempts from the backend otherwise.

Should the backend logs be showing anything from the frontend connection?

frontend console log:
```
69-9dd8c3df154f914b.js:1 No embedding models available
GET http://localhost:3000/library?_rsc=acgkz 404 (Not Found)
```

**Comment by ItzCrazyKns** [comment]
Seems like there are no models pulled by Ollama

**Comment by ItzCrazyKns** [comment]
I would recommend everyone to updating Perplexica to the latest version that I pushed today, it has a fix for a bug that might be causing issues here

**Comment by BarthPaleologue** [comment]
The latest version works! Thanks a lot

**Comment by delfireinoso** [comment]
I confirm that the last build is fine with Ollama running independently on Mac M2 Sonoma

**Comment by digitalw00t** [comment]
Did a git pull, then docker  compose up, I added curl to the back end, and it hits the ollama page saying it's running. I'm running the open-webui for ollama, pointing to the same end point, and I get a model list.

**Comment by NityaSG** [comment]
For me its running on my local windows but not on deployed Ubuntu VM

**Comment by gladly-hyphenated-21** [comment]
I am also having the same issue, also on the most recent pull.

**Comment by alexanderccc** [comment]
Same issue, I'm also trying to run it in Kubernetes on my local network, seems backend isn't receiving any requests from the client or the frontend itself, no errors in backend or frontend. Here's the traffic flow in Hubble after a fresh deployment and accessing the UI (backend app is not present)

![Screenshot from 2024-05-05 21-17-19](https://github.com/ItzCrazyKns/Perplexica/assets/48724148/a19d99f4-e130-4708-944b-093a6ba582ae)

Errors from console

![Screenshot from 2024-05-05 21-16-21](https://github.com/ItzCrazyKns/Perplexica/assets/48724148/5868fb29-2028-4060-abe1-09117105a8e8)

Logging/console errors aren't that descriptive, will look more into my setup next weekend, but any information would be useful

**Comment by ItzCrazyKns** [comment]
> Same issue, I'm also trying to run it in Kubernetes on my local network, seems backend isn't receiving any requests from the client or the frontend itself, no errors in backend or frontend. Here's the traffic flow in Hubble after a fresh deployment and accessing the UI (backend app is not present)
> 
> ![Screenshot from 2024-05-05 21-17-19](https://private-user-images.githubusercontent.com/48724148/328062692-a19d99f4-e130-4708-944b-093a6ba582ae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5NjU5NzEsIm5iZiI6MTcxNDk2NTY3MSwicGF0aCI6Ii80ODcyNDE0OC8zMjgwNjI2OTItYTE5ZDk5ZjQtZTEzMC00NzA4LTk0NGItMDkzYTZiYTU4MmFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA2VDAzMjExMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc1YjIwZGMyZjZlYWExMDBkMDQ0OWIzYzZiYTJlMGZmMTYyN2NjN2E0MWQzZmFmNzE0YmNhNTQ1YmExZTcyZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DOIw_1G5qxs1HTq37Rfh8oPH4kPNOvY6XHK_wCQPNNQ)
> 
> Errors from console
> 
> ![Screenshot from 2024-05-05 21-16-21](https://private-user-images.githubusercontent.com/48724148/328062740-5868fb29-2028-4060-abe1-09117105a8e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5NjU5NzEsIm5iZiI6MTcxNDk2NTY3MSwicGF0aCI6Ii80ODcyNDE0OC8zMjgwNjI3NDAtNTg2OGZiMjktMjAyOC00MDYwLWFiZTEtMDkxMTcxMDVhOGU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA2VDAzMjExMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2YTk0YWY0MmMzZjI5YTgzMTA2NzViNGQxN2RkNWVlZWI2ZWIxZTBlY2NmMWU3YjMzNGZkZGM0NWQxYWUwMzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Jv6sbFYHV9OKJLx25-8Rsud9UkNWYqCOBLptO1cDPjU)
> 
> Logging/console errors aren't that descriptive, will look more into my setup next weekend, but any information would be useful

The API URL seems invalid `http://domain/undefined` it should be `http://domain/api`

**Comment by ItzCrazyKns** [comment]
If the backend says the following:

```
perplexica-backend-1 | error: Error loading Ollama models: TypeError: fetch failed
perplexica-backend-1 | error: Error loading Ollama embeddings: TypeError: fetch failed
```
It suggests that the backend is not able to connect to the API of Ollama. You can try changing the API URL to the following:

On Windows: `http://host.docker.internal:11434`
On Mac: `http://host.docker.internal:11434`
On Linux: `http://private_ip_of_the_computer:11434`

**Comment by DaslavCl** [comment]
It's not just a ollama issue.. openAI too

**Comment by ItzCrazyKns** [comment]
Please provide more description

**Comment by NityaSG** [comment]
> Please provide more description

I have added groq and openai api keys in the config.toml (no Olama) 
When I am composing the docker image in my computer (windows) it is working completely fine . (however i observed that it worked only when i selected groq from the ui settings) 

I created an ubuntu VM (in gcp) and did all the standard procedures for setting up a docker container. It was deployed, and the searxng at port 4000 is working perfectly fine. 

However at port 3000 the UI and the  the settings bar shows loading forever
![Screenshot 2024-05-06 101052](https://github.com/ItzCrazyKns/Perplexica/assets/76156577/48f5ef43-e97a-4378-97d1-da4d3ac2ad79)

![Screenshot 2024-05-06 101259](https://github.com/ItzCrazyKns/Perplexica/assets/76156577/7d6837e8-2526-4805-8b1e-0ecb6fb8fd58)

**Comment by Yeqishen** [comment]
I tried deploying the project today on ubuntu, I used openai, but still got the same result. The page just keeps loading.

**Comment by ItzCrazyKns** [comment]
I everyone, unfortunately we do not provide support for the deployment of Peprlexica on some sort of VM or server. Our support is limited to its installation. All I can do is to tell you how you can fix this. You need to change the IP in the docker compose file to the IP of the server where you've deployed it, additionally make sure that the container is able to access Ollama (if you're using it)

**Comment by ItzCrazyKns** [comment]
If you are not able to see the model selector dropdown or the embedding model selector then update to the latest version. It has a patch for it.

**Comment by NityaSG** [comment]
<img width="413" alt="Screenshot 2024-05-06 161939" src="https://github.com/ItzCrazyKns/Perplexica/assets/76156577/8efd0ebb-693f-406c-abbf-f26799545bfb">
maybe something im doing wrong in setup

**Comment by ItzCrazyKns** [comment]
You are accessing it from your network. It is intended to access it from localhost, we do not provide support to set it up for network. We provide support for its installation locally

**Comment by NityaSG** [comment]
For the people who are deploying on Docker container and hosting it in a VM but getting endless loading : 
1. in Docker-compose.yaml
 `perplexica-frontend:
    build:
      context: .
      dockerfile: app.dockerfile
      args:
        - NEXT_PUBLIC_API_URL=http://127.0.0.1:3001/api
        - NEXT_PUBLIC_WS_URL=ws://127.0.0.1:3001`

change the 127.0.0.1 to your external ip address

2. Similarly in searxng-settings.yaml  change 
` bind_address: '127.0.0.1'` to your external ip address


3. If you get CORS error : open ui/components/ChatWindow.tsx and add cors header like this : 
` if (!chatModel || !chatModelProvider) {
          const chatModelProviders = await fetch(
            `${process.env.NEXT_PUBLIC_API_URL}/models`,{
              method: 'GET',
              **headers: {
                'Access-Control-Allow-Origin': '*',
              },**
            }
          ).then(async (res) => (await res.json())['providers']);`

**Comment by alexanderccc** [comment]
> For the people who are deploying on Docker container and hosting it in a VM but getting endless loading :
> 
>     1. in Docker-compose.yaml
>        `perplexica-frontend: build: context: . dockerfile: app.dockerfile args: - NEXT_PUBLIC_API_URL=http://127.0.0.1:3001/api - NEXT_PUBLIC_WS_URL=ws://127.0.0.1:3001`
> 
> 
> change the 127.0.0.1 to your external ip address
> 
>     2. Similarly in searxng-settings.yaml  change
>        ` bind_address: '127.0.0.1'` to your external ip address
> 
>     3. If you get CORS error : open ui/components/ChatWindow.tsx and add cors header like this :
>        `if (!chatModel || !chatModelProvider) { const chatModelProviders = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/models`,{ method: 'GET', **headers: { 'Access-Control-Allow-Origin': '*', },** } ).then(async (res) => (await res.json())['providers']);`

Yeah, this was my issue, was setting the apis as part of the env definition for the frontend instead of defining them at build. A runtime configuration for the frontend would be great to resolve this sort of issue. One thing to note, if running encrypted, make sure to set the endpoints as `https://` and `wss://` 
I've handled the CORS issues at the reverse proxy

**Comment by zhuozhiyongde** [comment]
> Yeah, this was my issue, was setting the apis as part of the env definition for the frontend instead of defining them at build. A runtime configuration for the frontend would be great to resolve this sort of issue. One thing to note, if running encrypted, make sure to set the endpoints as `https://` and `wss://` I've handled the CORS issues at the reverse proxy

@alexanderccc hey, could you please share you config? like Nginx config and `docker-compose.yaml`?

I also tried to config CORS, but it not work.

Here is my config of backend:

```nginx
location ^~ / {
    proxy_pass http://127.0.0.1:31338; 
    proxy_set_header Host $host; 
    proxy_set_header X-Real-IP $remote_addr; 
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for; 
    proxy_set_header REMOTE-HOST $remote_addr; 
    proxy_set_header Upgrade $http_upgrade; 
    proxy_set_header Connection "upgrade"; 
    proxy_set_header X-Forwarded-Proto $scheme; 
    proxy_http_version 1.1; 
    add_header X-Cache $upstream_cache_status; 
    add_header Strict-Transport-Security "max-age=31536000"; 
    
    # CORS
    add_header 'Access-Control-Allow-Origin' '*'; 
    add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
    add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range';
    add_header 'Access-Control-Expose-Headers' 'Content-Length,Content-Range';
}
```

and `docker-compose.yaml`:

```yaml
services:
  searxng:
    image: docker.io/searxng/searxng:latest
    volumes:
      - ./searxng:/etc/searxng:rw
    ports:
      - 31336:8080
    networks:
      - perplexica-network

  perplexica-backend:
    build:
      context: .
      dockerfile: backend.dockerfile
      args:
        - SEARXNG_API_URL=http://searxng:8080
    depends_on:
      - searxng
    ports:
      - 31338:31338
    networks:
      - perplexica-network

  perplexica-frontend:
    build:
      context: .
      dockerfile: app.dockerfile
      args:
        - NEXT_PUBLIC_API_URL=https://apidomain.of.myself/api
        - NEXT_PUBLIC_WS_URL=wss://apidomain.of.myself
        - PORT=31337
    depends_on:
      - perplexica-backend
    ports:
      - 31337:31337
    networks:
      - perplexica-network

networks:
  perplexica-network:
```

Both the backend and frontend utilize a reverse proxy.

**Comment by zhuozhiyongde** [comment]
Do I also need to replace the address/port of searXNG with an external HTTPS domain? If so, could you please tell me how to configure it? Thank you!

**Comment by NityaSG** [comment]
> Do I also need to replace the address/port of searXNG with an external HTTPS domain? If so, could you please tell me how to configure it? Thank you!

Yes,  in searxng-settings.yaml change bind_address: '127.0.0.1' to your external ip address

**Comment by alexanderccc** [comment]
I didn't have to touch the searxng config, I don't think it needs to be exposed as the backend uses it and not the frontend, so, as long as in the backend config (config.toml) the endpoint for it is defined correctly, you should be good.

For CORS I'm using Kong as an ingress controller which has a built-in plugin for handling it.

In your config for nginx you need to handle the CORS preflight check as part of an OPTIONS call, this post from Stackoverflow might help you with that setup

https://stackoverflow.com/questions/45986631/how-to-enable-cors-in-nginx-proxy-server

Also, be aware on the APIs that you're exposing, I think you would need to do an exact match on / for the frontend and the backend would need to be exposed for wss and the /api path, but I might be wrong, to keep it simple I exposed them with 2 different hosts (perplexica and perplexica-be) both on / with Prefix

**Comment by zhuozhiyongde** [comment]
> I didn't have to touch the searxng config, I don't think it needs to be exposed as the backend uses it and not the frontend, so, as long as in the backend config (config.toml) the endpoint for it is defined correctly, you should be good.
> 
> For CORS I'm using Kong as an ingress controller which has a built-in plugin for handling it.
> 
> In your config for nginx you need to handle the CORS preflight check as part of an OPTIONS call, this post from Stackoverflow might help you with that setup
> 
> https://stackoverflow.com/questions/45986631/how-to-enable-cors-in-nginx-proxy-server
> 
> Also, be aware on the APIs that you're exposing, I think you would need to do an exact match on / for the frontend and the backend would need to be exposed for wss and the /api path, but I might be wrong, to keep it simple I exposed them with 2 different hosts (perplexica and perplexica-be) both on / with Prefix

According to the solution you mentioned, I did not modify searXNG. Instead, I moved the CORS-related configuration from the reverse proxy to the server block, but it still didn't work. When I click on this `wss://` link, my backend returns a 502 Bad Gateway page. I would like to ask if there are any other configurations that need to be changed? Thank you for your reply.

**Comment by ItzCrazyKns** [comment]
Hi everyone, yesterday there were some issues related to networking those are fixed now so a few of the issues should be resolved. We unfortunately do not provide support to connect Perplexica to a domain but rather install it so we cannot help you in this

**Comment by alexanderccc** [comment]
I wouldn't say that there's no way, there is always a way to do something :sweat_smile:, but there is a difference between the project supporting it or not, I imagine @ItzCrazyKns wants to spend time actually developing the thing rather than providing support for running it remotely. 

I for one have been able to run this on my home lab on kubernetes connected to ollama and I'm using it from my pc/laptop or phone, and honestly, it's been pretty awesome to run queries from my phone and test it out that way :grin: 

Maybe we can close this off and open a separate issue to document the various setups for running it remotely?

**Comment by zhuozhiyongde** [comment]
> I wouldn't say that there's no way, there is always a way to do something 😅, but there is a difference between the project supporting it or not, I imagine @ItzCrazyKns wants to spend time actually developing the thing rather than providing support for running it remotely.
> 
> I for one have been able to run this on my home lab on kubernetes connected to ollama and I'm using it from my pc/laptop or phone, and honestly, it's been pretty awesome to run queries from my phone and test it out that way 😁
> 
> Maybe we can close this off and open a separate issue to document the various setups for running it remotely?

I apologize for my inappropriate expression (and possibly disturbing emails). I did not intend to waste everyone's time; it's just that after genuinely trying and still failing, I felt quite discouraged and thus became a bit anxious. I think it would indeed be better to open a new issue.

**Comment by ItzCrazyKns** [comment]
Yes, I've created a new discussion here: https://github.com/ItzCrazyKns/Perplexica/discussions/111, here everyone can ask questions about deployment remotely and the community can provide support and help them overcome their issues, I myself would love to help them when I am free.

**Comment by Wadera** [comment]
After few tries and errors I've managed to make it working and wanted to share my setup ;)

## Ports: 

I strongly suggest to leave defaults ports of `3000` for frontend and `3001` for backend. I've tried to adjust it in `docker-compose.yml` and `config.toml` for backend but then app doesn't want to work. Maybe this ports are hardcoded somewhere? Maybe my initial build cash it?  I'm not sure, but if you struggle - leave them as defaults. 

## docker-compose.yml: 

If  you want to make it working over domain - you need to [fallow documentation](https://github.com/ItzCrazyKns/Perplexica/blob/master/docs/installation/NETWORKING.md) and adjust it to IP or domain name. That's my working config:

```
  perplexica-frontend:
    build:
      context: .
      dockerfile: app.dockerfile
      args:
        #- NEXT_PUBLIC_API_URL=http://192.168.40.10:3001/api                 # working over IP
        #- NEXT_PUBLIC_WS_URL=ws://192.168.40.10:3001                        # working over IP
        - NEXT_PUBLIC_API_URL=https://perpexica.mydomain.tdl/api              # working over domain
        - NEXT_PUBLIC_WS_URL=wss://perpexica.mydomain.tdl/wss                 # working over domain
    depends_on:
      - perplexica-backend
    ports:
      - 3000:3000
    networks:
      - perplexica-network
    restart: unless-stopped
```

Where `perpexica.mydomain.tdl` is my domain with working https protocol and `192.168.40.10` is my docker VM IPv4 (in my case in internal subnet, as I'm just running it on my home-lab environment). 

If you decide to use IP - then you can visit your IP with port to display frontend, so example: http://192.168.40.10:3000 

## Reverse proxy:

If you decide that you want to run it over https and friendly domain - you need to deploy reverse proxy. I'm using [Nginx Proxy Manager](https://github.com/NginxProxyManager/nginx-proxy-manager), but you can deploy or use any other proxy you want. 

To avoid issues with CORS I strongly suggest to proxy everything under single domain - both frontend and backend. In my case it would be `perpexica.mydomain.tdl` 

### That's GUI setup:

Details:
![image](https://github.com/user-attachments/assets/c6270b96-12e5-4090-b69c-d73b382afb02)

Custom locations:
![image](https://github.com/user-attachments/assets/cefc7eb4-015c-446b-b26d-071e1c220dc1)

SSL:
![image](https://github.com/user-attachments/assets/370f42db-e234-40ba-b5fa-5dfb7e73fa31)

### proxy code
And that's code generated by npm if you want to deploy it with different tools:

```
# ------------------------------------------------------------
# perpexica.mydomain.tdl
# ------------------------------------------------------------

map $scheme $hsts_header {
    https   "max-age=63072000; preload";
}

server {
  set $forward_scheme http;
  set $server         "192.168.40.10";
  set $port           3000;

  listen 80;
listen [::]:80;
listen 443 ssl http2;
listen [::]:443 ssl http2;

  server_name perpexica.mydomain.tdl;

  # Let's Encrypt SSL
  include conf.d/include/letsencrypt-acme-challenge.conf;
  include conf.d/include/ssl-ciphers.conf;
  ssl_certificate /etc/letsencrypt/live/npm-68/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/npm-68/privkey.pem;

    # Force SSL
    include conf.d/include/force-ssl.conf;

proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $http_connection;
proxy_http_version 1.1;

  access_log /data/logs/proxy-host-66_access.log proxy;
  error_log /data/logs/proxy-host-66_error.log warn;

  location /api {
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For    $remote_addr;
    proxy_set_header X-Real-IP          $remote_addr;
    proxy_pass       http://192.168.40.10:3001/api;

    # Authorization
    auth_basic            "Authorization required";
    auth_basic_user_file  /data/access/1;

    proxy_set_header Authorization "";


    # Access Rules: 4 total
    allow 192.168.2.0/24;
    allow 192.168.40.0/24;
    allow 172.0.0.1/8;
    allow 192.168.1.0/24;
    deny all;

    # Access checks must...
    satisfy any;

    # Force SSL
    include conf.d/include/force-ssl.conf;

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;

  }
  location /wss {
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For    $remote_addr;
    proxy_set_header X-Real-IP          $remote_addr;
    proxy_pass       http://192.168.40.10:3001;

    # Authorization
    auth_basic            "Authorization required";
    auth_basic_user_file  /data/access/1;
    proxy_set_header Authorization "";

    # Access Rules: 4 total
    allow 192.168.2.0/24;
    allow 192.168.40.0/24;
    allow 172.0.0.1/8;
    allow 192.168.1.0/24;
    deny all;

    # Access checks must...
    satisfy any;

    # Force SSL
    include conf.d/include/force-ssl.conf;

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;
  }

  location / {
    # Authorization
    auth_basic            "Authorization required";
    auth_basic_user_file  /data/access/1;

    proxy_set_header Authorization "";

    # Access Rules: 4 total
    allow 192.168.2.0/24;
    allow 192.168.40.0/24;
    allow 172.0.0.1/8;
    allow 192.168.1.0/24;
    deny all;

    # Access checks must...
    satisfy any;

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;

    # Proxy!
    include conf.d/include/proxy.conf;
  }

  # Custom
  include /data/nginx/custom/server_proxy[.]conf;
}
```

## Security

As you hopefully noticed - I'm using my Access List which would allow to visit my instance from local network by my family members or over VPN to my home. 

Be aware that all your search history and API keys are exposed under cog setting page and potentially visible in requests (I didn't review code that deeply). 

That's just one of the reason why [ItzCrazyKns](https://github.com/ItzCrazyKns) advising to not expose this service to public at this point of development. 

Until proper login panel, user management and security would be improved - I strongly advise to use it locally.

## Final words: 

Thank you for bringing this project to live! You are all doing brilliant work with it's development. I hope that my notes would be useful to others ;)
